src/commands/inputs.rs

```diff
@@ -0,0 +1,484 @@
+//! Implementation of the `input` command.
+
+use std::collections::HashSet;
+
+use anyhow::Result;
+use anyhow::anyhow;
+use anyhow::bail;
+use clap::Parser;
+use serde_json::Map;
+use serde_json::Value;
+use serde_yaml_ng;
+use wdl::analysis::Document;
+use wdl::analysis::types::CallKind;
+use wdl::ast::AstNode;
+use wdl::ast::AstToken;
+use wdl::ast::v1::Decl;
+use wdl::ast::v1::Expr;
+use wdl::ast::v1::InputSection;
+use wdl::ast::v1::LiteralExpr;
+use wdl::ast::v1::TaskDefinition;
+use wdl::ast::v1::Type;
+use wdl::cli::Analysis;
+use wdl::cli::analysis::Source;
+
+/// Arguments for the `input` subcommand.
+#[derive(Parser, Debug)]
+pub struct Args {
+    /// A source WDL file or URL.
+    #[arg(value_name = "PATH or URL")]
+    pub source: Source,
+
+    /// The name of the task or workflow for which to generate inputs.
+    #[clap(short, long, value_name = "NAME")]
+    pub name: Option<String>,
+
+    /// Show inputs with non-literal default values and non-required inputs.
+    #[arg(long)]
+    pub show_expressions: bool,
+
+    /// Include inputs with default values.
+    #[arg(long)]
+    pub include_defaults: bool,
```

Comment:
It's not clear to me just from these docstrings how these flags interact.
How would I show optional inputs without a default? Those are semantically equivalent to a defaulted input with `= None`, so they should appear with `--include-defaults`, but that should be made explicitly clear in the docstring

Reply:
With the way it is currently written, `--show-expressions` does nothing on its own.

Reply:
It actually makes little sense to me how it is currently written. Outputs aren't consistently ordered between runs. Also this `"rnaseq_standard.bam_to_fastqs.fqlint.disable_validator_codes": "Array[String]",` is output for `--include-defaults. It explicitly does not include the `default`, which is only included if you also supply `--show-expressions`: `"rnaseq_standard.bam_to_fastqs.fqlint.disable_validator_codes": "Array[String] (default = [])",`

Reply:
I think the best behavior might be:
```suggestion
    /// Show inputs with literal default values and non-required inputs.
    #[arg(long)]
    pub show_literal_defaults: bool,

    /// Include inputs with complex expressions for defaults. Ignored if `--show-literal-defaults` not specified.
    #[arg(long)]
    pub show_expressions: bool,
```
?

Reply:
@claymcleod - It looks like you took a much different approach to what these arguments were meant to be. Can you clarify your intent with them? It looks like your implementation of `show_expressions` only controls whether the complex expression component is included in the output. My original implementation omitted the entire input if its value was a complex expression.

Reply:
IMO that should be:
```
"rnaseq_standard.bam_to_fastqs.fqlint.disable_validator_codes": []
```

Reply:
@claymcleod - Also I'm curious if you know immediately why the ordering of the output varies between runs.

Reply:
> IMO that should be:
> 
> ```
> "rnaseq_standard.bam_to_fastqs.fqlint.disable_validator_codes": []
> ```

I'm fairly certain that was the previous behavior, but I haven't rolled back the changes to check.

Reply:
I'm tempted to shelve this until @claymcleod weighs in. It appears he has a significantly different vision for this subcommand than the one we outlined previously. I'd like to hear his reasons before attempting to recapitulate the prior behavior.

Reply:
For the output ordering, running the following command with [the commit I started with](https://github.com/claymcleod/sprocket/commit/03db56387e164ea3cb5cc5ef9efbc051f484e515) yields inputs in different orders across runs.

```
sprocket input https://raw.githubusercontent.com/stjudecloud/workflows/refs/heads/main/workflows/rnaseq/rnaseq-standard.wdl --show-expressions --nested-inputs
```

So, though I also did not check it, it seems that this behavior was prior before the PR. This makes sense, because `serde_json::Map` does not preserve the order of inserts unless you tell it to with the `preserve_order` feature.

---

For the other pieces, admittedly I did totally interpret what these flags were meant to be differently. What I was aiming for was:

- `--includes-defaults` only modulates whether input parameters with a default value are shown.
- `--show-expressions` plays off of the idea of showing the more complicated expression (as defined by `decl.expr()`) when an easy conversion to a value is not known.

Notably, the confusing bit might be that _some_ expressions are converted to literals automatically if they are straightforward enough, regardless of the value of `--show-expressions`. That being said, the underlying point is that I appear to have taken a stronger separation of concerns approach regarding what "including defaults" and "including expressions" means (it is true that the `--show-expressions` flag has no effect without `--include-defaults` because required inputs never have an expression i.e. they are unbound).

If I understand right, the two are you are advocating for a more practical approach, whereby

- `--show-defaults` is not just about whether or not to display a default input parameter but also around showing the default value directly rather than annotating its type.
- for `--show-expressions`... I'm not actually sure.

Take the code I referred to above that I started with. If you run this command:

```bash
sprocket input https://raw.githubusercontent.com/stjudecloud/workflows/refs/heads/main/workflows/rnaseq/rnaseq-standard.wdl --show-expressions
```

You get this output.

```json
{
  "rnaseq_standard.bam": "File",
  "rnaseq_standard.gtf": "File",
  "rnaseq_standard.star_db": "File",
  "rnaseq_standard.xenocp_aligner": "star",
  "rnaseq_standard.strandedness": "",
  "rnaseq_standard.mark_duplicates": false,
  "rnaseq_standard.cleanse_xenograft": false,
  "rnaseq_standard.validate_input": true,
  "rnaseq_standard.use_all_cores": false,
  "rnaseq_standard.subsample_n_reads": -1
}
```

But if you run it with `--show-expressions`:

```bash
sprocket input https://raw.githubusercontent.com/stjudecloud/workflows/refs/heads/main/workflows/rnaseq/rnaseq-standard.wdl --show-expressions
```

You get this output.

```json
{
  "rnaseq_standard.bam": "File",
  "rnaseq_standard.gtf": "File",
  "rnaseq_standard.star_db": "File",
  "rnaseq_standard.contaminant_db": null,
  "rnaseq_standard.prefix": "String (default = basename(bam, \".bam\"))",
  "rnaseq_standard.xenocp_aligner": "star",
  "rnaseq_standard.strandedness": "",
  "rnaseq_standard.mark_duplicates": false,
  "rnaseq_standard.cleanse_xenograft": false,
  "rnaseq_standard.validate_input": true,
  "rnaseq_standard.use_all_cores": false,
  "rnaseq_standard.subsample_n_reads": -1
}
```

The difference here is the following:

```txt
4a5,6
>   "rnaseq_standard.contaminant_db": null,
>   "rnaseq_standard.prefix": "String (default = basename(bam, \".bam\"))",
```

A couple of questions come out of that.

1. Doesn't `rnaseq_standard.contaminant_db` have a default value of `null` regardless of if `--show-expressions` is shown? Why is it only showing up for `--show-expressions`? I don't see the relation between that showing up with the flag at all. 

From the spec:

> An optional declaration has a default initialization of `None`, which indicates that it is undefined. An optional declaration may be initialized to any literal or expression of the correct type, including the special None value.

2. Perhaps similarly, `rnaseq_standard.prefix` _does_ have a default value—it's just that it's a complex expression. Given that it has a default value of type `String`, I would expect to see it in the list by default with a value of `"String"` since this implementation shows default values without `--hide-defaults`.

Reply:
Here's the behavior I want to see _as a user_:

Without any flags, I would want just the required inputs with a type annotation as the value.

I'd like to modify that behavior with a flag that will fill in literal defaults (i.e., I don't have to edit these values to get this file to run, so they should not have type annotations. This would also include `null` for optional inputs).

Then there would be a "show me everything that can be tweaked" flag, which includes things that are calculated at runtime based on other values (complex expressions) that can't be rendered in JSON. These should have a type annotation and the runtime expression as a string.

Lastly, nested inputs should be viewable with a flag.

Reply:
The above is the optimal behavior of the command IMO in terms of maximizing usefulness, but I would be interested in hearing any alternatives.

As for the ordering of the JSON, I think the best behavior is to probably put "you need to edit this field" values at the top of the file (so required inputs and "complex expression" values that are type annotated), but I'm ambivalent about sub-ordering beyond that (assuming it's consistent across re-runs).

Reply:
> I'd like to modify that behavior with a flag that will fill in literal defaults (i.e., I don't have to edit these values to get this file to run, so they should not have type annotations. This would also include null for optional inputs).

After some more thought, I'd argue that should probably be the default behavior, as it's the most useful. I imagine that most times I personally use this command, I would be specifying this flag. Many of our WDL workflows as well as external workflows are really meant to have some of the defaults overridden (e.g. read2 FASTQs which are defaulted to an empty list).

That said, only outputting the required inputs is a sane default behavior and would be ok if someone strongly disagrees with me.

Reply:
Yeah I agree with all of the above. If Andrew agrees with it, I think the flags should be tweaked in this direction. I do think it's a good idea to go ahead and print literals without requiring a flag when we can.